bobtimus/src/lib.rs

```diff
@@ -1,8 +1,173 @@
+            [(outpoint, txout)] => {
+                use hmac::{Hmac, Mac, NewMac};
+                use sha2::Sha256;
+
+                let master_blinding_key = self.elementsd.dumpmasterblindingkey().await?;
+                let master_blinding_key = hex::decode(master_blinding_key)?;
+
+                let mut mac = Hmac::<Sha256>::new_varkey(&master_blinding_key)
+                    .expect("HMAC can take key of any size");
+                mac.update(txout.script_pubkey().as_bytes());
+
+                let result = mac.finalize();
+                let input_blinding_sk = SecretKey::from_slice(&result.into_bytes())?;
+
+                ((*outpoint, txout.clone()), input_blinding_sk)
+            }
+            [] => bail!("found no inputs"),
+            _ => bail!("TODO: Support multiple inputs per party"),
+        };
+
+        let redeem_address = self.elementsd.getnewaddress().await?;
+        let change_address = self.elementsd.getnewaddress().await?;
+
+        let protocol_state = Bob0::new(
+            payload.btc_amount.0,
+            usdt_amount.0,
+            input,
+            input_blinding_sk,
+            self.btc_asset_id,
+            redeem_address,
+            change_address,
+        );
+
+        let protocol_state = protocol_state.interpret(&mut self.rng, payload.protocol_msg)?;
+        let tx = self
+            .elementsd
+            .sign_raw_transaction(protocol_state.unsigned_transaction())
+            .await?;
+
+        Ok(tx)
+    }
+}
+
+#[async_trait]
+pub trait LatestRate {
+    async fn latest_rate(&self) -> Result<Rate>;
+}
+
+/// Rate in USDt per bitcoin.
 #[derive(Debug, Clone, Copy)]
-pub struct Rate(pub u64);
+pub struct Rate(pub Decimal);
 
+#[derive(Clone, Copy, PartialEq)]
+pub struct LiquidUsdt(Amount);
+
+impl TryFrom<f64> for LiquidUsdt {
+    type Error = anyhow::Error;
+
+    fn try_from(value: f64) -> Result<Self> {
+        Ok(LiquidUsdt(Amount::from_btc(value)?))
+    }
+}
+
+impl From<LiquidUsdt> for Amount {
+    fn from(from: LiquidUsdt) -> Self {
+        from.0
+    }
+}
+
+impl std::fmt::Debug for LiquidUsdt {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "LiquidUsdt({} dollars)", self.0.as_btc())
+    }
+}
+
+#[derive(Debug, Clone, Copy, Deserialize)]
+pub struct BitcoinAmount(
+    #[serde(with = "::elements_fun::bitcoin::util::amount::serde::as_sat")] Amount,
+);
+
+impl Mul<BitcoinAmount> for Rate {
+    type Output = LiquidUsdt;
+
+    fn mul(self, rhs: BitcoinAmount) -> Self::Output {
+        let amount = self.0.mul(Decimal::from(rhs.0.as_sat())).to_u64().unwrap();
+
+        LiquidUsdt(Amount::from_sat(amount))
+    }
+}
+
+impl From<Amount> for BitcoinAmount {
+    fn from(amount: Amount) -> Self {
+        Self(amount)
+    }
+}
+
+impl From<BitcoinAmount> for Amount {
+    fn from(amount: BitcoinAmount) -> Self {
+        amount.0
+    }
+}
+
+// TODO: This controls how we serialise the Rate. It would be better
```

Comment:
:+1:

Reply:
I fixed it by deriving serialization and forcing serialization of the rate before passing to `warp::sse::data()`, instead of letting it use the `Display` implementation of the type by default, which is a weird choice.

Reply:
Much better to just use `warp::see::json()` as you suggested below.